[extension][ecsobserver] Init ECS service discovery using docker label

pingleig opened the pull request:
**Description:** <Describe what has changed.>

This is the first (divided) implementation PR for #1395. In order to reduce the size of the pr, the feature is incomplete.

- mock api client is used, no actual aws api calls
- Only docker label based filter (matcher) is included (others are just place holders)
- Extension impl under `ecsobserver` is excluded, it's a thin wrapper.

Those excluded logic are tested in other branches, so it works with real API end to end (eventually).

~~The package is located under internal/aws/ecssd~~ (relocated to extension because currently there is no other aws related component reusing it)

**Link to tracking Issue:** <Issue number if applicable>

- Feature request #1395 
- Config PR (merged) https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2377

**Testing:** <Describe what testing was performed and which tests were added.>

unit test

**Documentation:** <Describe the documentation added.>

No new doc, see https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2377 for existing doc

## Known issues

- customise `job` in service discovery output will cause scrape error, https://github.com/open-telemetry/opentelemetry-collector/issues/575 so a new config is added and default to `prometheus_job`

pingleig committed:
extension: Init ECS service discovery in internal package

This is the first (divided) implementation PR for #1395.
In order to reduce the size of the pr, the feature is incomplete.

- mock api client is used, no actual aws api calls
- Only docker label based filter/matcher is included
- Extension impl under `ecsobserver` is excluded

Those excluded logic are tested in other branches, so it works with real
API end to end (eventually).

The package is located under internal/aws/ecssd because

- it can be used by other aws specific packages
- we may extract it out to a standalone repo in the future
  as it can run as a binary and support other prometheus
  compatiable collectors as a sidecar

pingleig commented:
I guess the bot should have assigned it to @jrcamp

pingleig committed:
aws: ecssd Remove unused code

- ignore *.actual.yaml and only check in *.expected.yaml

jrcamp commented:
@pingleig overall looks good. Can somebody at AWS review first? Quite a lot of code.

pingleig commented:
@jrcamp thanks! I just asked in internal channel, there should be some follow up PRs, those should be smaller because most interface and large struct definitions are in this PR.

mxiamxia commented on internal/aws/ecssd/filter.go:
```diff
@@ -0,0 +1,90 @@
+	"go.uber.org/zap"
+)
+
+type TaskFilter struct {
+	logger   *zap.Logger
+	matchers map[MatcherType][]Matcher
+}
+
+type TaskFilterOptions struct {
+	Logger   *zap.Logger
+	Matchers map[MatcherType][]Matcher
+}
```
want to learn why we can't just use one structure if it is exactly the same -_-

pingleig replied:
We can, I just removed the options struct and pass the parameter directly in the new func (\w/)

pingleig committed:
extension: Move ecssd into extension/observer/ecsobeserver

There is no plan to reuse it in other aws related components

pingleig committed:
extension: Init go mod and makefile for ecssd

pingleig committed:
internal: Revert go mod for internal/aws

pingleig committed:
extension: ecssd implement extension interface

pingleig commented on internal/aws/ecssd/docker_label.go:
```diff
@@ -0,0 +1,111 @@
+func (d *DockerLabelConfig) Init() error {
+	// It's possible to support it in the future, but for now just fail at config,
+	// so user don't need to wonder which port is used in the exported target.
+	if len(d.MetricsPorts) != 0 {
```
I don't want to split the shared config because docker label is just one of three matchers, though its behaviour is pretty different from service and task based matcher so we could give it another struct like `ExportConfigWithoutPorts`. I plan to delay the change for next pr when service & task are added.

pingleig replied:
Actually I think since we are going to have https://github.com/open-telemetry/opentelemetry-collector/pull/2802 it's better to put the validate logic here instead of removing it from struct. Because if user use `metrics_ports` for docker label and it's not in the struct, it will throw unmarshal error without telling why it is not there while it is supported for other filters. Though in the long term I think docker label should become part of common export config as well, you can discover based on service name and pick port based on docker label in task definition, we are just trying to be consistent with cloudwatch agent's behaviour here at this moment.

pingleig commented on internal/aws/ecssd/config.go:
```diff
@@ -0,0 +1,91 @@
+
+// ExampleConfig returns an example instance that matches testdata/config_example.yaml.
+// It can be used to validate if the struct tags like mapstructure, yaml are working properly.
+func ExampleConfig() Config {
```
This is actually not the default config used by components,  `createDefaultConfig` is added after I moved the code to extension folder.

pingleig commented on internal/aws/ecssd/config.go:
```diff
@@ -0,0 +1,91 @@
+	// ClusterRegion is the target ECS cluster's AWS region.
+	ClusterRegion string `mapstructure:"cluster_region" yaml:"cluster_region"`
+	// RefreshInterval determines how frequency at which the observer
+	// needs to poll for collecting information about new processes.
```
the code is moved from internal/aws to extension.

pingleig committed:
extension: ecssd Add docker label to default config

Also fix go fmt problems ...

pingleig commented:
oops, windows test failed because of `\r\n` and `\n` when comparing generated file https://app.circleci.com/pipelines/github/open-telemetry/opentelemetry-collector-contrib/11473/workflows/8898ac3b-3231-431e-a457-625ca5ca6442/jobs/97574 I will grab a windows pc and test it (if I don't end up playing games)

pingleig committed:
extension: ecssd Repalce \r\n with \n for windows test

The expected file is converted to \r\n during git checkout, however the
generated file is still using \n. This only happens on circle ci's
windows environemnt. See https://circleci.com/blog/circleci-config-teardown-how-we-write-our-circleci-config-at-circleci/#main:~:text=Line%20endings

mxiamxia reviewed (commented):
LGTM in general. pls address the comments. Thanks!

pingleig committed:
extension: ecssd Make default config values internal

- Fix typo
- Add more comment on how the matcher merge logic works

pingleig commented on extension/observer/ecsobserver/filter.go:
```diff
@@ -0,0 +1,85 @@
+			for _, container := range res.Containers {
+				matchedTasks[container.TaskIndex] = true
+				task := tasks[container.TaskIndex]
+				task.AddMatchedContainer(container)
```
Add the comment. We are updating those tasks in place because we still need those attached information (task def, ec2 etc.) when we are exporting the target. We could create another struct with almost identical fields and make shallow copy but I feel the current way is simpler as the matched target is the only thing we are modifying, and it's easier to write unit tests.

pingleig commented:
@mxiamxia addressed comments in https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2734/commits/5d200bddcb48ffdadacd87db0ebc3d274693fe1b and @jrcamp PTAL (btw: the code grows a bit bigger because the extension interface is implemented compared w/ last time you look at it)

mxiamxia reviewed (approved):
LGTM! Thanks!

jrcamp commented:
@pingleig Can we find a way to split this up more? I just don't have the cycles to digest such a large PR all in one.

pingleig commented:
@jrcamp it's doable, but in that case I don't think I can meet the requirement for test coverage because they will not work end to end (even in unit test). If that's ok I can create a new pr for just structs.

pingleig commented:
@jrcamp created https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/2939

pingleig commented:
Close it since half of the PR is split into #2939 and I plan create a new one instead of rebase on this branch.
